[#167] Removal of unnecessary catch blocks

Eclipse-Dominator opened the pull request:
Fixes #167 

Alternative approach: #186 (Where functionality of the catch blocks is restored)

---

Following issue #167, all IOExceptions are already being handled within `JsonUtil#readJsonFile()`, therefore additional catch blocks as well as throwing of IOExceptions are no longer needed.

This PR moves handling of conversion errors as well as regular IOExceptions to within `JsonUtil#readJsonFile()`. As a result of this change, it is no longer appropriate to call the exception thrown `DataConversionException`. 

Thus, the above-mentioned exception will be renamed to `DataLoadingException`. Affected interfaces and code caused by this change will then be updated.

damithc reviewed (commented):
Quick feedback: The renaming of the exception class deserves its own commit?

damithc commented on src/main/java/seedu/address/commons/exceptions/DataConversionException.java:
```diff
@@ -1,7 +1,7 @@
 
 /**
- * Represents an error during conversion of data from one format to another
+ * Represents an error during loading of files from the file system.
```
I think this belongs in the other commit, together with the name change.

damithc replied:
Strictly speaking, the correct order is to tweak the exception first, and then remove the catch blocks. The latter is more justifiable after the exception has been tweaked.

damithc commented on src/main/java/seedu/address/commons/util/JsonUtil.java:
```diff
@@ -49,14 +49,16 @@ static <T> T deserializeObjectFromJsonFile(Path jsonFile, Class<T> classOfObject
     /**
-     * Returns the Json object from the given file or {@code Optional.empty()} object if the file is not found.
-     * If any values are missing from the file, default values will be used, as long as the file is a valid json file.
+     * Returns the Json object from the given file or {@code Optional.empty()} object if the file is
```
Not strictly related to this PR but as long as we are touching these lines, perhaps use `JSON` or `json` instead of `Json`?

damithc commented on src/main/java/seedu/address/commons/util/JsonUtil.java:
```diff
@@ -49,14 +49,16 @@ static <T> T deserializeObjectFromJsonFile(Path jsonFile, Class<T> classOfObject
      * @param filePath cannot be null.
      * @param classOfObjectToDeserialize Json file has to correspond to the structure in the class given here.
-     * @throws DataConversionException if the file format is not as expected.
+     * @throws DataConversionException if the loading of file failed
```
Not sure if `loading of the file` will be understood as both 'reading the file content and parsing the content'. Perhaps use `loading of JSON data from the file`?

Eclipse-Dominator replied:
For this version of PR, the handling of trying to read file content and parsing the content is both handled within `JsonUtil` so in a sense the new renamed exception represents both 'reading the file content and parsing the content'.

damithc commented on src/main/java/seedu/address/storage/JsonUserPrefsStorage.java:
```diff
@@ -32,8 +32,9 @@ public Optional<UserPrefs> readUserPrefs() throws DataConversionException {
+     *
      * @param prefsFilePath location of the data. Cannot be null.
-     * @throws DataConversionException if the file format is not as expected.
+     * @throws DataConversionException if the loading of the preference file fails.
```
This is phrasing is slightly different from the one used in `readAddressBook`. Don't vary things unless there is a good reason.

damithc commented on src/main/java/seedu/address/storage/JsonAddressBookStorage.java:
```diff
@@ -40,7 +40,7 @@ public Optional<ReadOnlyAddressBook> readAddressBook() throws DataConversionExce
      *
      * @param filePath location of the data. Cannot be null.
-     * @throws DataConversionException if the file is not in the correct format.
+     * @throws DataConversionException if the loading of the data fails.
```
fails -> failed?

damithc commented on src/test/java/seedu/address/commons/util/ConfigUtilTest.java:
```diff
@@ -68,7 +68,8 @@ private Config getTypicalConfig() {
 
-    private Optional<Config> read(String configFileInTestDataFolder) throws DataConversionException {
+    private Optional<Config> read(String configFileInTestDataFolder)
+            throws DataConversionException {
```
This change is not needed?

damithc commented on src/test/java/seedu/address/storage/JsonAddressBookStorageTest.java:
```diff
@@ -47,17 +47,20 @@ public void read_missingFile_emptyResult() throws Exception {
     public void read_notJsonFormat_exceptionThrown() {
-        assertThrows(DataConversionException.class, () -> readAddressBook("notJsonFormatAddressBook.json"));
+        assertThrows(DataConversionException.class,
+                () -> readAddressBook("notJsonFormatAddressBook.json"));
```
Revert unrelated changes.

damithc reviewed (changes requested):
Some suggestions added.

damithc commented on src/main/java/seedu/address/MainApp.java:
```diff
@@ -10,7 +10,7 @@
 import seedu.address.commons.core.Config;
```
Commit message can be shorter:
```
Rename DataConversoinException -> ConfigLoadingException

DataConversoinException is thrown to indicate any problem
encountered while reading a JSON config file, not just problems
specific to data conversion.

Let's rename it as ConfigLoadingException to make the name
more representative of its purpose.
```

damithc commented on src/main/java/seedu/address/commons/exceptions/ConfigLoadingException.java:
```diff
@@ -3,8 +3,8 @@
 /**
  * Represents an error during conversion of data from one format to another
```
`Represents an error during loading of data from a config file.`

damithc reviewed (changes requested):
The first commit should be just the renaming. It is a mechanical change and requires minimal review. The only other thing that can be go into that commit is the tweaking of the javadoc comment of that class so that the comment matches the new class name.
Once that is out of the way, the next commit can focus on behavioral tweaks.

damithc commented on src/main/java/seedu/address/storage/StorageManager.java:
```diff
@@ -54,12 +54,12 @@ public Path getAddressBookFilePath() {
 
     @Override
-    public Optional<ReadOnlyAddressBook> readAddressBook() throws DataConversionException, IOException {
+    public Optional<ReadOnlyAddressBook> readAddressBook() throws ConfigLoadingException, IOException {
```
Wait, this doesn't look right. I think we have to name it `DataLoadingException` instead.

Eclipse-Dominator committed:
Rename DataConversionException -> DataLoadingException

DataConversionException is thrown to indicate any problem
encountered while reading a JSON file, not just problems
specific to data conversion.

Let's rename it as DataLoadingException to make the name
more representative of its purpose.

damithc reviewed (changes requested):
Mostly OK. I suggest the following:
* Commit 1: Remove everything else except those strictly related to renaming the exception. It should look almost like a mechanical find-and-replace refactoring. 
* Commit 2: commit message: No need to mention 'In the previous commit'. The previous commit speaks for itself. Here, mention that IOException is unnecessary and is removed, and logs/docs are adjusted accordingly.

damithc commented on src/main/java/seedu/address/MainApp.java:
```diff
@@ -125,7 +121,7 @@ protected Config initConfig(Path configFilePath) {
             initializedConfig = configOptional.orElse(new Config());
         } catch (DataLoadingException e) {
-            logger.warning("Config file at " + configFilePathUsed + " is not in the correct format."
+            logger.warning("Config file at " + configFilePathUsed + " could not be read."
```
Should this be `could not be loaded.` instead?

damithc commented on src/main/java/seedu/address/storage/AddressBookStorage.java:
```diff
@@ -19,16 +19,16 @@ public interface AddressBookStorage {
-     *   Returns {@code Optional.empty()} if storage file is not found.
+     * Returns {@code Optional.empty()} if storage file is not found.
+     *
      * @throws DataLoadingException if the data in storage is not in the expected format.
```
Doesn't this need updating as well?

Eclipse-Dominator replied:
I missed out this when doing the rebasing to shift the documentation change to the 2nd commit.

damithc commented on src/main/java/seedu/address/MainApp.java:
```diff
@@ -84,13 +84,9 @@ private Model initModelManager(Storage storage, ReadOnlyUserPrefs userPrefs) {
             }
```
Suggested alternative commit message:
```
JsonUtil#readJsonFile() method throws a DataLoadingException when it
detects an IOException.  However, some client code of that method are 
written as if it throws an IOException.

Let's,
1. remove 'throws IOExceptions' clause from the relevant client methods
   in Storage, AddressBookStorage, and UserPrefsStorage.
2. remove corresponding catch blocks.
3. tweak relevant javadocs and log message to match.
```

damithc reviewed (commented):
Commit 1: Clean and easy to review. Nice!
Commit 2: Added a few cosmetic suggestions.

damithc commented:
@Eclipse-Dominator Can you update the PR description to match the current state of the PR? After that, I'll ask others for reviews.

Eclipse-Dominator committed:
Refactor client code handling IOException

JsonUtil#readJsonFile() method throws a DataLoadingException when it
detects an IOException.  However, some client code of that method are
written as if it throws an IOException.

Let's,
1. remove 'throws IOExceptions' clause from the relevant client methods
   in Storage, AddressBookStorage, and UserPrefsStorage.
2. remove corresponding catch blocks.
3. tweak relevant javadocs and log message to match.

damithc commented:
@se-edu/tech-team-level1 for your review ...

SPWwj reviewed (approved):
LGTM! Now, the MainApp.java has become more streamlined, and the exceptions are more specific. Excellent work!

damithc commented:
Thanks @Eclipse-Dominator for the fix, and @SPWwj for the review.
